src/wrappers/themis/wasm/src/secure_cell_context_imprint.js

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2019 Cossack Labs Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file
+ * Themis Secure Cell (Context Imprint mode).
+ */
+
+const libthemis = require('./libthemis.js')
+const errors = require('./themis_error.js')
+const utils = require('./utils.js')
+
+const subsystem = 'SecureCellContextImprint'
+
+const ThemisError = errors.ThemisError
+const ThemisErrorCode = errors.ThemisErrorCode
+
+module.exports = class SecureCellContextImprint {
+    constructor(masterKey) {
+        masterKey = utils.coerceToBytes(masterKey)
+        if (masterKey.length == 0) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'master key must be not empty')
+        }
+        this.masterKey = masterKey
+    }
+
+    encrypt(message, context) {
+        message = utils.coerceToBytes(message)
+        if (message.length == 0) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'message must be not empty')
+        }
+
+        // Other Secure Cell kinds have optional context, so the users are likely to omit
+        // the context here as well. Let's produce a more helpful error message instead
+        // of an error about 'undefined' being invalid byte buffer type.
+        if (context === undefined) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'SecureCellContextImprint requires context')
+        }
+
+        context = utils.coerceToBytes(context)
+        if (context.length == 0) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'context must be not empty')
+        }
+
+        let status
+        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
+        let result_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
```

Comment:
`i32` - is it 32-bytes long? will it be compatible with x64 systems? (imaginary case if someone decides to encrypt 5gb of data, results will be > 2^32 bytes)

Reply:
Yes, i32 is LLVM type of signed 32-bit integers. Emscripten is 32-bit platform as JavaScript does not support 64-bit integers. WebAssembly has both 32-bit and 64-bit flavors, but we'll be using only 32-bit one.

As for compatibility, this is more general question of interoperability between 32-bit and 64-bit Themis. WasmThemis is the 32-bit one. It will not be able to encrypt or decrypt more than 4 GB of data ever.

Note that it's `i32` (signed), not `u32` (unsigned). So currently we're limited to 2 GB. For some reason Emscripten does not like `u32` type which should be correct representation of unsigned `size_t`. We'll have to get back to this at some point.